test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineRetransform/RedefineRetransform.java

```diff
@@ -0,0 +1,274 @@
+        int hookVer = getClassBytesVersion(hookClassBytes);
+        if (hookVer != ver) {
+            throw new RuntimeException("CLFH got unexpected version: "  + hookVer
+                    + " (expected " + ver + ")");
+        }
+    }
+
+    // Retransforms TestClass instrumenting (from ClassFileLoadHook) with 'ver' class bytes (if >= 0).
+    // Verifies that class bytes passed to ClassFileLoadHook have correct version (expectedVer).
+    static void retransform(int ver, int expectedVer) {
+        byte[] classBytes = getClassBytes(ver);
+        byte[] hookClassBytes = nRetransform(TestClass.class, classBytes);
+        int hookVer = getClassBytesVersion(hookClassBytes);
+        if (hookVer != expectedVer) {
+            throw new RuntimeException("CLFH got unexpected version: "  + hookVer
+                    + " (expected " + expectedVer + ")");
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        int testCase;
+        try {
+            testCase= Integer.valueOf(args[0]);
+        } catch (Exception ex) {
+            throw new RuntimeException("Single numeric argument expected", ex);
+        }
+        init();
+        switch (testCase) {
+        case 1:
+            test("Redefine-Retransform-Retransform", () -> {
+                redefine(1);        // cached class bytes are not set
+                retransform(2, 1);  // sets cached class bytes to ver 1
+                retransform(3, 1);  // uses existing cache
+            });
+            break;
+
+        case 2:
+            test("Redefine-Retransform-Redefine-Redefine", () -> {
+                redefine(1);        // cached class bytes are not set
+                retransform(2, 1);  // sets cached class bytes to ver 1
+                redefine(3);        // resets cached class bytes
+                redefine(4);        // cached class bytes are not set
+            });
+            break;
+
+        case 3:
+            test("Redefine-Retransform-Redefine-Retransform", () -> {
+                redefine(1);        // cached class bytes are not set
+                retransform(2, 1);  // sets cached class bytes to ver 1
+                redefine(3);        // resets cached class bytes
+                retransform(4, 3);  // sets cached class bytes to ver 3
+            });
+            break;
+
+        case 4:
+            test("Retransform-Redefine-Retransform-Retransform", () -> {
+                retransform(1, 0);  // sets cached class bytes to ver 0 (initially loaded)
+                redefine(2);        // resets cached class bytes
+                retransform(3, 2);  // sets cached class bytes to ver 2
+                retransform(4, 2);  // uses existing cache
+            });
+            break;
+
+        case 5:
+            test("Redefine-Retransform-Redefine-Retransform with CFLH", () -> {
+                redefine(1, 5);     // CFLH sets cached class bytes to ver 1
```

Comment:
Test cases 1-4 are from the bug description.
I added test cases 5 & 6 to verify additional code paths - they are the same as 3 & 4, but in RedefineClasses we provide new class bytes in CFLH.
I.e. in cases 3 and 4 after RedefineClasses classes have no cached bytes and class bytes are reconstituted in the subsequent Retransform;
In case 5 and 6 cache_bytes buffer is created during RedefineClasses,  RetransformClasses use existing cache.

Reply:
Thanks for the explanation.